src/Microsoft.AspNetCore.OData/Extensions/Endpoint/ODataEndpointLinkGenerator.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+#if !NETSTANDARD2_0
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNet.OData.Extensions
+{
+    /// <summary>
+    /// Defines a contract to generate absolute and related URIs based on OData endpoint routing.
+    /// </summary>
+    internal class ODataEndpointLinkGenerator : LinkGenerator
+    {
+        private static readonly string _escapedHashMark = Uri.EscapeDataString("#");
+        private static readonly string _escapedQuestionMark = Uri.EscapeDataString("?");
+
+        private LinkGenerator _innerGenerator;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ODataEndpointLinkGenerator"/> class.
+        /// </summary>
+        /// <param name="generator">The inner Link generator</param>
+        public ODataEndpointLinkGenerator(LinkGenerator generator)
+        {
+            _innerGenerator = generator;
```

Comment:
@xuzhg Is there a chance of this constructor being called with a null value for the `LinkGenerator` parameter? That would most certain precipitate chaos when the overridden methods that reference `_innerGenerator` are called

Reply:
Theoretically, no. But, let me add the verification code.

---
In reply to: [384365505](https://github.com/OData/WebApi/pull/2035#discussion_r384365505) [](ancestors = 384365505)